### PR TITLE
Wrap attribute value getter in try-catch block

### LINKED ISSUE #21: IE8 attribute change detection fails if DOM object has a property with failing toString() method
Open the following demo in IE8 and observe the error http://output.jsbin.com/dinovu/

The problem here is in the fact that IE8 considers properties as attributes and calls toString() when you  are trying to evaluate attribute value

### PR DESCRIPTION
**Pull request by mnaoumov**
Fixes #21

### CODE REVIEW

**Comment by megawac** [comment]
Thanks for investigating, I'll test and profile this over the weekend.

**Comment by megawac** [comment]
I just tested this and this fix doesn't prevent the error. It should also be noted that calling `.value` on such a node doesn't actually throw an exception. Instead it seems to `console.error`

**Comment by mnaoumov** [comment]
Yeah, you're right it was an inaccurate testing

**Comment by megawac** [comment]
Seems to get redirected directly to `window.onerror` which is a PITA

**Comment by mnaoumov** [comment]
It seems that it is even worse, as I tried to override

```js
console.error = function() {}
window.onerror = function() { return false; }
```

and the error is still shown

**Comment by mnaoumov** [comment]
@megawac I've made a new fix which works now. Please check http://output.jsbin.com/jimoca/

File: MutationObserver.js [hunk 1]
```diff
@@ -199,6 +199,21 @@ window.MutationObserver = window.MutationObserver || (function(undefined) {
...
     }
 
+    /**
+     * Gets an attribute value ignoring errors
+     *
+     * @param {Attr} attr
+     * @return {String} an attribute value or null in case of errors
+     */
+    function getAttributeValueSafe(attr) {
+        try {
```
Comment (megawac) [inline MutationObserver.js]: Is the `try`...`catch` still necessary?

**Comment by megawac** [comment]
Mind doing a `git cherry-pick 7944242` and I'll run sauce on this branch

**Comment by megawac** [comment]
Currently the performance hit is to great to even consider (2-13 times slower in IE10) for this edge case

File: MutationObserver.js (hunk 1 above)
Comment (mnaoumov) [inline MutationObserver.js]: Yes, try-catch is needed because we need to capture toSting() failure

**Comment by mnaoumov** [comment]
@megawac I cherry-picked the suggested commit and build fails now. I'm going to investigate it shortly. Regarding the performance issue, this hack is required only in IE<9 because later on attributes are not mixed with properties. See http://javascript.info/tutorial/attributes-and-custom-properties#the-fun-with-old-ies. So I consider to add browser detection

**Comment by mnaoumov** [comment]
@megawac I've implemented a check if browser distinguishes between attributes and properties to avoid unnecessary overhead

**Comment by megawac** [comment]
Alright, I'll reprofile this. I am also going to toy around some techniques to detect and omit non standard attributes

**Comment by mnaoumov** [comment]
@megawac, I thought about that as well. Naturally you project should watch only against DOM attributes rather than object properties. I could not find any documented way to do that. So far, I had an idea to get element's outerHTML and parse real DOM attribute names from there via regex. This is required only for IE<9

**Comment by megawac** [comment]
Can you explain the use case of this change? Why are you using custom
attributes which can't be `toString`ed

**Comment by mnaoumov** [comment]
@megawac , the real world example of this problem, which I actually discovered at my project, https://github.com/mnaoumov/jhtmlarea/issues/7 , you can the demo for this issue in IE8 at http://output.jsbin.com/pezice

I fixed the issue in toString for jHtmlArea project but I think it worth fixing it in your project as well.

**Comment by megawac** [comment]
In your case was the editor begin working after initiation? What I mean to
say was the error actually just a warning message or did it actually break
something

**Comment by mnaoumov** [comment]
It is just a warning message. But in case if your Developer Tools are open and it has **Break on Exceptions** setting, the debugger stops on the toString() method and pollutes to the console

**Comment by megawac** [comment]
I'm going to profile this to see what the impact is, however, but I'm begining to think this should just be discussed in documentation

**Comment by megawac** [comment]
Right now, I'm thinking an inline comment on the function so developers who have **Break On\*** will be referred to this issue and a note in the readme. Thoughts?

**Comment by mnaoumov** [comment]
Yeah, that's the way to go in case if working this around is too costy. Right now I am working on a completely different implementation. I want to capture only real DOM attributes and skip the ones that came from custom properties. In this case we would be able to call **attr.value** safely. Moreover, it may slightly improve performance for cases when DOM element has a few attributes and many properties. So I think it worth implementing anyway regardless of this bug itself

**Comment by megawac** [comment]
Yeah I was planning to look into the same thing and would def give that a
+1 if it works well

**Comment by mnaoumov** [comment]
@megawac , I've managed to implement the desired behavior. At first I started with parsing outerHTML but found that doing this with regex too error-prone. Then I invented a very simple cloning approach. It may decrease performance for IE<9 but all other browsers should not be affected.

**Comment by mnaoumov** [comment]
@megawac Did you check this new implementation?

**Comment by megawac** [comment]
Not yet, haven't been on windows

**Comment by megawac** [comment]
Yeah, this is way too much of a perf hit to be worthwhile IMO. I'm just going to document it as its only warning anyway

**Pull request closed without merging**